static/js/Editor.jsx

```diff
@@ -2546,18 +2547,152 @@ const BlockButton = ({format, icon}) => {
     )
 }
 
+const EditorSaveStateIndicator = ({ state }) => {
+    const localize = (inputStr) => Sefaria._(inputStr, "EditorSaveIndicator");
+    const stateToIcon = {
+      "connectionLost": "/static/icons/new_editor_saving/cloud-off-rounded.svg",
+      "userUnauthenticated": "/static/icons/new_editor_saving/person-off.svg",
+      "saving": "/static/icons/new_editor_saving/directory-sync-rounded.svg",
+      "saved": "/static/icons/new_editor_saving/cloud-done-rounded.svg",
+      "unknownError": "/static/icons/new_editor_saving/cloud-off-rounded.svg"
+    };
+    const stateToTooltip = {
+    "saved": "Your sheet is saved to Sefaria",
+    "saving": "We are saving your changes to Sefaria",
+    "connectionLost": "No internet connection detected",
+    "userUnauthenticated": "You are not logged in to Sefaria",
+    "unknownError": "If this problem persists, please try again later and contact us at hello@sefaria.org"
+  };
+
+    const path = window.location.pathname + window.location.search;
+    const stateToMessage = {
+      "connectionLost": "Trying to connect…",
+      "userUnauthenticated": <>{localize("User Logged out")}. <a href={`/login?next=${path}`}>{localize("Log in")}</a></>,
+      "saving": "Saving...",
+      "saved": "Saved",
+      "unknownError": "Something went wrong. Try refreshing the page."
+    };
+    const loadedIcons = new Set();
+    useEffect(() => {
+      if (state === "connectionLost") {
+        return;
+      }
+      // Preload icons for all states if they are not already loaded.
+      Object.values(stateToIcon).forEach((src) => {
+        if (!loadedIcons.has(src)) {
+          const img = new Image();
+          img.src = src;
+          img.onload = () => loadedIcons.add(src);  // Mark as loaded after successful load
+        }
+      });
+    }, [state]);
+
+    const tooltip = stateToTooltip[state];
+
+    return (
+        <ToolTipped altText={localize(tooltip)} classes={`editorSaveStateIndicator tooltip-toggle ${state}`}>
+        {<img src={stateToIcon[state]} alt={localize(state)} />}
+        <span className="saveStateMessage">{localize(stateToMessage[state])}</span>
+        </ToolTipped>
+  );
+}
+function useUnsavedChangesWatcher(timeoutSeconds, unsavedChanges, savingState, setSavingState) {
+  const timeoutRef = useRef(null);
+
+  // if failed to save and no reason is given, transition into known error state after a timeout
+  useEffect(() => {
+    const hasUnsaved = unsavedChanges;
```

Comment:
what is the purpose of renaming this variable? I think it only creates confusion. Also, `hasUnsaved` implies the variable is a boolean but `unsavedChanges` is ambiguous. If it is a boolean, prefer `hasUnsaved`.